cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,594 @@
+            self.rules.bindings.push(binding.clone());
+            self.binding_map.insert(binding, id);
+            id
+        }
+    }
+
+    fn dedup_expr(&mut self, expr: Expr) -> ExprId {
+        if let Some(expr) = self.expr_map.get(&expr) {
+            *expr
+        } else {
+            let id = ExprId(self.rules.exprs.len().try_into().unwrap());
+            self.rules.exprs.push(expr.clone());
+            self.expr_map.insert(expr, id);
+            id
+        }
+    }
+
+    fn set_constraint(&mut self, input: Binding, constraint: Constraint) -> BindingId {
+        let input = self.dedup_binding(input);
+        if let Err(e) = self.current_rule.set_constraint(input, constraint) {
+            self.unreachable.push(e);
+        } else {
+            self.set_constraint_on_class(input, constraint);
+        }
+        input
+    }
+}
+
+impl sema::PatternVisitor for RuleSetBuilder {
+    type PatternId = Binding;
+
+    fn add_match_equal(&mut self, a: Binding, b: Binding, _ty: sema::TypeId) {
+        let a = self.dedup_binding(a);
+        let b = self.dedup_binding(b);
+        // If both bindings represent the same binding site, they're implicitly equal.
+        if a != b {
+            self.current_rule.equals.merge(a, b);
+            // If either binding had a constraint, copy that constraint to the whole equivalence
+            // class (including the other binding).
+            if let Some(constraint) = self.current_rule.get_constraint(a) {
+                self.set_constraint_on_class(a, constraint);
+            } else if let Some(constraint) = self.current_rule.get_constraint(b) {
+                self.set_constraint_on_class(b, constraint);
+            }
+        }
+    }
+
+    fn add_match_int(&mut self, input: Binding, _ty: sema::TypeId, int_val: i128) {
+        self.set_constraint(input, Constraint::ConstInt(int_val));
+    }
+
+    fn add_match_prim(&mut self, input: Binding, _ty: sema::TypeId, val: sema::Sym) {
+        self.set_constraint(input, Constraint::ConstPrim(val));
+    }
+
+    fn add_match_variant(
+        &mut self,
+        input: Binding,
+        input_ty: sema::TypeId,
+        arg_tys: &[sema::TypeId],
+        variant: sema::VariantId,
+    ) -> Vec<Binding> {
+        let fields = TupleIndex(arg_tys.len().try_into().unwrap());
+        let source = self.set_constraint(input, Constraint::Variant(fields, input_ty, variant));
+        (0..fields.0)
+            .map(TupleIndex)
+            .map(|field| Binding::Variant {
+                source,
+                variant,
+                field,
+            })
+            .collect()
+    }
+
+    fn add_extract(
+        &mut self,
+        input: Binding,
+        _input_ty: sema::TypeId,
+        output_tys: Vec<sema::TypeId>,
+        term: sema::TermId,
+        infallible: bool,
+        _multi: bool,
+    ) -> Vec<Binding> {
+        // External extractor invocations are expressions in Rust
```

Comment:
Add a comment here why convert to expr-land and dedup -- so that we strongly normalize external-extractor-of-a-given-binding-site

Reply:
Strongly normalizing is a red herring in this case. External extractor calls would be normalized whether we represent them as an `Expr` or as a `Binding` here. The entire reason for converting from pattern to expression and back around external extractor calls is because I designed this representation to closely resemble what we can easily implement with Rust as our codegen target, and you can't do a function call in a Rust pattern match. I've added comments in a variety of places which I hope would have clarified this for you if I'd written them before your code review.